src/adapter/src/catalog/open.rs

```diff
@@ -584,32 +574,37 @@ impl Catalog {
                 config,
             } in clusters
             {
-                let introspection_source_index_gids = txn.get_introspection_source_indexes(id);
+                let introspection_source_index_ids = txn.get_introspection_source_indexes(id);
 
                 let AllocatedBuiltinSystemIds {
                     all_builtins: all_indexes,
                     new_builtins: new_indexes,
                     ..
                 } = Catalog::allocate_system_ids(&mut txn, BUILTINS::logs().collect(), |log| {
-                    introspection_source_index_gids
+                    introspection_source_index_ids
                         .get(log.name)
                         .cloned()
                         // We migrate introspection sources later so we can hardcode the fingerprint as ""
-                        .map(|id| SystemObjectUniqueIdentifier {
+                        .map(|(id, _oid)| SystemObjectUniqueIdentifier {
                             id,
                             fingerprint: "".to_string(),
                         })
                 })?;
 
                 let new_indexes = new_indexes
                     .iter()
-                    .map(|(log, index_id)| IntrospectionSourceIndex {
-                        cluster_id: id,
-                        name: log.name.to_string(),
-                        index_id: *index_id,
-                    })
+                    .map(|(log, index_id)| (id, log.name.to_string(), *index_id))
                     .collect();
-                txn.set_introspection_source_indexes(new_indexes)?;
+                let new_indexes: BTreeMap<_, _> = txn.insert_introspection_source_indexes(new_indexes)?.into_iter().map(|introspection_source_index| (introspection_source_index.name, introspection_source_index.oid)).collect();
+
+                let all_indexes: Vec<_> = all_indexes.into_iter().map(|(log, id)| {
+                    // First look in existing indexes.
+                    let oid = introspection_source_index_ids.get(log.name).map(|(_id, oid)| oid);
+                    // Then look in new indexes.
+                    let oid = oid.or_else(|| new_indexes.get(log.name));
+                    let oid = oid.cloned().expect("log, {log:?}, with ID {id:?} must exist in one of the maps");
```

Comment:
I don't think you can do string formatting in a `.expect(...)`

Reply:
Fixed